SimPEG/ObjectiveFunction.py

```diff
@@ -0,0 +1,326 @@
+    debug = False
+
+    mapPair = Maps.IdentityMap  #: Base class of expected maps
+    _mapping = None  #: An IdentityMap instance.
+
+    _nP = None
+
+    def __init__(self, **kwargs):
+        Utils.setKwargs(self, **kwargs)
+
+    def __call__(self, x, **kwargs):
+        return self._eval(x, **kwargs)
+
+    @property
+    def nP(self):
+        if self._nP is not None:
+            return self._nP
+        if getattr(self, 'mapping', None) is not None:
+            return self.mapping.nP
+        return '*'
+
+    @property
+    def mapping(self):
+        if getattr(self, '_mapping') is None:
+            if getattr(self, '_nP') is not None:
+                self._mapping = self.mapPair(nP=self.nP)
+            else:
+                self._mapping = self.mapPair()
+        return self._mapping
+
+    @mapping.setter
+    def mapping(self, value):
+        assert issubclass(value, self.mapPair), (
+            'mapping must be an instance of a {}, not a {}'
+        ).format(self.mapPair, value.__class__.__name__)
+
+
+    @Utils.timeIt
+    def _eval(self, x, **kwargs):
+        raise NotImplementedError(
+            "The method _eval has not been implemented for {}".format(
+                self.__class__.__name__
+            )
+        )
+
+    @Utils.timeIt
+    def deriv(self, x, **kwargs):
+        raise NotImplementedError(
+            "The method deriv has not been implemented for {}".format(
+                self.__class__.__name__
+            )
+        )
+
+    @Utils.timeIt
+    def deriv2(self, x, **kwargs):
+        raise NotImplementedError(
+            "The method _deriv2 has not been implemented for {}".format(
+                self.__class__.__name__
+            )
+        )
+
+    def _test_deriv(self, x=None, num=4, plotIt=False, **kwargs):
+        print('Testing {0!s} Deriv'.format(self.__class__.__name__))
+        if x is None:
+            if self.nP == '*':
+                x = np.random.randn(np.random.randint(1e2, high=1e3))
+            else:
+                x = np.random.randn(self.nP)
+
+        return checkDerivative(
+            lambda m: [self(m), self.deriv(m)], x, num=num, plotIt=plotIt,
+            **kwargs
+        )
+
+    def _test_deriv2(self, x=None, num=4, plotIt=False, **kwargs):
+        print('Testing {0!s} Deriv2'.format(self.__class__.__name__))
+        if x is None:
+            if self.nP == '*':
+                x = np.random.randn(np.random.randint(1e2, high=1e3))
+            else:
+                x = np.random.randn(self.nP)
+
+        return checkDerivative(
+            lambda m: [self.deriv(m), self.deriv2(m)], x, num=num,
+            plotIt=plotIt, **kwargs
+        )
+
+    def test(self, x=None, num=4, plotIt=False, **kwargs):
+        deriv = self._test_deriv(x=x, num=num, **kwargs)
+        deriv2 = self._test_deriv2(x=x, num=num, plotIt=False, **kwargs)
+        return (deriv & deriv2)
+
+    __numpy_ufunc__ = True
+
+    def __add__(self, objfct2):
```

Comment:
adding these in a reasonable way is a bit subtle... so far, this is the best I have come up with

![image](https://cloud.githubusercontent.com/assets/6361812/22305681/479af890-e2f1-11e6-9e57-7178ca1945e9.png)